### DISCUSSION TITLE
Linking issues with the Scope class (potentially because of the Case Paths sub-dependency)

### DISCUSSION THREAD

**Discussion by amandaroa** [pointfreeco/swift-composable-architecture#2459] [General]
We are trying to use TCA with a new feature on an existing app. It seems that in our release/prod scheme I am having linking errors when Xcode tries to compile code that uses Case Paths, like the Scope class from TCA.  I have added both The Composable Architecture and CasePaths as dependencies using SPM. We are using this code in an Embedded Framework and get the warning that  "Linking against a dylib which is not safe for use in application extensions" (Wondering if this is the case for only proper Extensions or if an embedded framework counts as general application extension). But the main issue is the error I am seeing. Everything works fine locally with our dev environment but not when trying to run locally with our release scheme. 

The error I am seeing is "Undefined symbols for architecture arm64:". See image below for all of it. Are there any architecture limitations to this framework? Or any specific build settings I need to adjust? 

![Screenshot 2023-09-13 at 5 34 41 PM](https://github.com/pointfreeco/swift-case-paths/assets/35345278/83c9013a-bb98-4caa-88ac-cfda81ea1f55)

When I comment out out Scope declaration, it will build (but it breaks our implementation). So I have pinpointed it to this specific class:
`  
        Scope(state: \.radioButtonFeature, action: /Action.radioButtonFeature) {
            RadioButtonFeature()                
        }
`

It seems that we clear that error if we change "Build Active Architecture Only" to "Yes" but we have other unrelated errors come up when we do that so I'm hoping there is another way around this.

**Comment by stephencelis** [comment]
Thanks for starting a discussion around this! I wonder if it's possible to attach a reduced project that reproduces the issue? Maybe we'll be able to figure out that the problem is and workaround it that way.

**Reply by amandaroa** [reply]
My apologies, I did end up finding a work around. I am still not sure if there may be an adjustment needed from the package or if it is an deeper issue with how our dependencies are set up. I"ll drop in some context and a summary in case it is helpful or useful to you:


 We did have a bit of a unique pattern we were using where we had several local packages in an effort to modularize out code, one of which used The Composable Architecture. We put all our local packages in an embedded framework called SPMLibraries. We could then add SPMLibraries as a dependency on whatever target we wanted without having to add each local package individually to each target. 

A couple weeks ago, there was some linking issue a while ago also related to the lines of code I showed above, and to fix it I had to individually also add Case Paths as a dependency to our local package. This seemed odd to me but it got us moving forward. Meanwhile, I didn't realize that while it worked for our debug scheme, it did not work for our release scheme. 

Fast forward to today, removed the local package that used TCA from the embedded framework (SPMLibraries) and added it to our main target. Luckily we didn't have to share this local package with the other targets like the others. That resolved all the issues.I was also able at that point to remove declaration of the Case Path dependency and found that the sub-dependencies for TCA were getting imported correctly now.

I am not sure why it was having the issue when it was in the Embedded Framework. I did make a quick test project and recreated the structure of handling the dependencies and it worked just fine and could compile with those same lines of code. So it may just be a hiccup in our setup. But if not, hope some of this context can be helpful to you if anyone else starts experiencing a similar issue.

**Reply by jshier** [reply]
I've noticed the same thing in Xcode 15, including the RC. Between the app target and a single feature target I have to keep manually linking the various TCA-exported dependencies in order for linking to complete correctly. It seems something has changed in regards to exported frameworks in Xcode 15. It's not the new linker, as setting the old linker doesn't fix the issue (but does give better error diagnostics for it). My exploration of TCA is [here](https://github.com/jshier/TCAExploration). (Xcode 15 only.) You can see that the various targets are manually linked to TCA and its dependencies in the build phases. If you remove any of them the build will fail to link.

**Reply by amandaroa** [reply]
Good to know it's not just our project! I experienced it in Xcode 14.3.1 and indeed the error messages were not very clear. Your observation that it is multi-target issue lines up exactly with what I was experiencing. Thanks for sharing!

**Reply by dazy1030** [reply]
Hello. I am having the same problem and found this discussion. Have you already found a solution? In my project, it is difficult to eliminate the TCA dependency on the module side, so I am looking for a solution that keeps the TCA dependent structure both in the app and in the module.

I have tried several solutions and found that the following two points will result in a link error. (although the cause needs to be explored further)
1. both the app and the module have TCA in place.
2. use DependencyKey on the app side.

I have a simple project (link_error branch) that causes the problem. Thanks.

https://github.com/dazy1030/TCASample/tree/link_error
